### PR TITLE
Fix "Prompt is too long" session deadlock (#93)

### LINKED ISSUE #93: error: Prompt is too long
When trying to reverse the API of a popular grocery website store locator, after running several minutes it gives this error: 
error: Prompt is too long

I think it would be nice to just cut the prompt down to a manageable size should it get very long instead of stopping.. or perhaps prompt the user to adjust it manually.

### PR DESCRIPTION
**Pull request by kalil0321**
Fixes #93.

## Root cause

Claude Code's auto-compaction leaves less headroom than a single max-size tool result — Read caps at ~25k tokens per call. While the agent pages through a large HAR (a grocery-site capture with `record_har_content="embed"` is easily 50–200 MB), one big read landing between the compact check and the next API request jumps straight past the hard limit. At that point the conversation itself exceeds the model's limit, so every subsequent request — **including the auto-compaction request** — fails with the same 400, and the session is permanently stuck. Parallel tool calls widen the overshoot (compaction can only run between API requests, never mid-batch), and base64/minified JSON tokenizes worse than prose, so HAR content maximizes the risk. Same failure class as upstream anthropics/claude-code#46348 / #25867 / #7594.

Per the [Claude Code env-vars docs](https://code.claude.com/docs/en/env-vars), proactive compaction (compacting *before* the limit) is not even guaranteed to be active: for local sessions on newer models, auto-compaction otherwise waits until the conversation reaches the model's context limit — exactly the deadlock-prone behavior.

This is why the error is hard to reproduce: it needs a heavy site, a session that creeps near the threshold, and a large tool-result batch landing in the narrow band between the compact trigger and the hard limit.

## Changes

- **Configure proactive auto-compaction per the documented recipe** for all Claude SDK sessions (engineer, agent mode, collector) via `build_sdk_env()` in `utils.py`: `CLAUDE_CODE_AUTO_COMPACT_WINDOW` (a large value the CLI caps at the model's real context window, enabling proactive compaction everywhere) plus `CLAUDE_AUTOCOMPACT_PCT_OVERRIDE=85` (≥30k tokens of headroom on a 200k window — more than one max-size Read). Both are documented at code.claude.com/docs/en/env-vars, and user-set values win (the SDK merges our dict on top of `os.environ`; we only add a key when it's absent). No prompt changes.
- **Graceful failure when it still happens** (parallel batches can exceed the headroom): context-overflow errors (`prompt is too long`, `conversation too long`, `exceed context limit`) are detected in `_process_streaming_response`, a clear explanation is printed (progress is saved on disk; start a new run for the same target to continue; shorter browsing sessions keep the HAR smaller), and the follow-up loop ends instead of offering more turns on a dead session where every message would fail identically.

Note: the collector session previously passed no `env` at all; it now uses the same `build_sdk_env()` as the other SDK sessions, which also clears `CLAUDECODE` there for consistency (verified against claude-agent-sdk 0.1.48: `env` is merged on top of `os.environ`, not a replacement).

## Testing

- New tests: `build_sdk_env()` defaults and per-variable user-override precedence (including explicit-empty), overflow-error classification, overflow flag + help on error results, follow-up loop exit after overflow, per-run overflow-flag reset, and env wiring into `ClaudeAgentOptions`.
- Full suite: 835 passed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Hsp5pTY71Pr8YwQ3ed5UZJ

<h3>Greptile Summary</h3>

This PR prevents Claude sessions from becoming stuck after context-window overflow. The main changes are:

- Lowers the default auto-compaction threshold to 85% while preserving user overrides.
- Applies the shared SDK environment to engineer, agent, and collector sessions.
- Detects context-overflow responses and ends unusable follow-up sessions.
- Resets overflow state when an engineer instance starts a new run.
- Adds tests for environment handling and overflow behavior.

<h3>Confidence Score: 5/5</h3>

This looks safe to merge.

The reused-instance overflow state is reset before each new SDK session. Explicit empty and non-empty auto-compaction overrides remain intact. The collector environment is merged with the process environment by the pinned SDK.

No blocking issues remain in the changed code.

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- A proof was produced for the posted P1 finding and linked to the corresponding review comment.
- Contract-validation was run for the Claude session overflow, showing the before state with CLAUDECODE and two follow-up prompts, and the after state with threshold 85, one follow-up prompt, one help call, and an overflow state.
- A focused test suite for the after state was executed, and all 111 tests passed.
- A remaining failure was identified in the missing-env assertion from the folder-sdk-env after log.
- Artifacts summarizing the validation were prepared, including the before/after/after-focused logs and related Python artifacts.

<a href="https://app.greptile.com/trex/runs/15419349/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/reverse_api/engineer.py | Adds overflow detection, user guidance, per-run state reset, SDK environment setup, and follow-up termination. |
| src/reverse_api/auto_engineer.py | Applies the shared SDK environment and overflow lifecycle handling to Claude agent mode. |
| src/reverse_api/collector.py | Applies the shared SDK environment to collector sessions. |
| src/reverse_api/utils.py | Adds SDK environment construction and context-overflow error classification. |
| tests/test_engineer.py | Covers overflow detection, follow-up termination, and SDK environment wiring. |
| tests/test_utils.py | Covers default, explicit, and empty environment overrides plus overflow message classification. |

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. General comment 

   <a href="#">P1</a> **Folder-name Claude SDK session does not receive build_sdk_env**

   - **Bug**
     - `_generate_folder_name_async()` creates a real Claude Agent SDK session for generating folder names, but its `ClaudeAgentOptions` contains only `allowed_tools`, `permission_mode`, and `model`. Consequently this session neither clears `CLAUDECODE` nor receives the default auto-compaction threshold of 85, contradicting the requirement that all Claude SDK session constructors receive `build_sdk_env()`.
   - **Cause**
     - The constructor at `src/reverse_api/utils.py:160` was omitted when env wiring was added to engineer, auto-engineer, and collector constructors.
   - **Fix**
     - Pass `env=build_sdk_env()` to the `ClaudeAgentOptions` call in `_generate_folder_name_async()` and add a focused mocked test asserting its exact env dictionary, including explicit-empty override preservation.

   <sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>

</details>

<sub>Reviews (2): Last reviewed commit: ["Follow documented recipe for proactive a..."](https://github.com/kalil0321/reverse-api-engineer/commit/2fff56fe06386679a46a27c786480ad6e42264e5) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=46325724)</sub>

### CODE REVIEW

File: src/reverse_api/collector.py [hunk 1]
```diff
@@ -112,6 +112,7 @@ async def _agent_loop(self) -> dict[str, Any] | None:
             ],
             cwd=str(self._collected_dir),
             model=self.model,
+            env=build_sdk_env(),
```
[the remark this answers is not in the document]
Comment (kalil0321) [reply src/reverse_api/collector.py:115]: Verified against the pinned SDK (claude-agent-sdk 0.1.48): `_internal/transport/subprocess_cli.py` builds the child environment as `{**os.environ, **(options.env or {}), ...}`, so `env` is merged on top of the full process environment, not used as a replacement. The engineer and agent-mode sessions have always passed a two-key `env={"CLAUDECODE": ""}` dict through this same path — replace semantics would have stripped `PATH` and broken CLI spawning there long ago. Also relevant: `build_sdk_env()`'s override-respecting behavior depends on this merge contract (omitting the key lets a user-set `CLAUDE_AUTOCOMPACT_PCT_OVERRIDE` pass through from `os.environ`), and that behavior is covered by tests. No change needed.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by kalil0321** [comment]
@greptile review
